### PR TITLE
Feature/63 recommendationsystem

### PR DESCRIPTION
**Pull request by niraymak**
## Description

This image shows the made functionality.

C3

We added ElasticSearch to our project to serve as a recommendation system. ElasticSearch uses a NoSQL database so we had to update the database on start. This has to be done manually by using the /export endpoint in the ProjectController on staging / production.
CRUD operations on projects (and project likes) will be updated to ElasticSearch automatically.

Project recommendations are based on collaborative filtering. This means the system tries to find 'similar users'. Similar users are found based on equal project likes. The user with the most equal project likes is the most similar user. The system will start recommending projects of that users if you don't like them yet. If that user does not have any projects left to recommend it will start looking in the second best similar user etc. etc.

Project recommendations can be retrieved using the /projectrecommendations/{amount} UserController. {amount} is the amount of project you would like to retrieve.

## Type of change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] This change requires a documentation update

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I did not update API Controllers, if I did, I added/changed Postman tests
-   [x] I have performed a self-review of my own code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [x] I have made corresponding changes to the documentation
-   [x] I have added tests that prove my fix is effective or that my feature works
-   [x] New and existing unit tests pass locally with my changes
-   [x] I updated the changelog with an end-user readable description
-   [x] I assigned this pull request to the correct project board to update the sprint board

## Steps to Test or Reproduce

There are three ways described below to review the created feature. 

1. As you need RabbitMQ and ElasticSearch we recommend using Docker to test. Clone code and use command: docker-compose up --build in the root folder.
2. Since u are probably testing in development the database will automatically be migrated. Please keep in mind that the data is randomly seeded.
3. Use Swagger -> Log in -> Use UserController /projectrecommendations to retrieve recommendations. Please keep in mind that the behaviour is unexpected since the data is seeded. 

1. Review new Postman collection / environment ( in folder in repo: /Postman/ElasticSearch )
2. Run in Docker using command: docker-compose up --build
3. Run tests

1. Run unit tests

## Link to issue

Closes: #63

### CODE REVIEW

File: ElasticSynchronizer/Helperclasses/JsonParser.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+using ElasticSynchronizer.Models;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace ElasticSynchronizer.Helperclasses
```
Comment (Brend-Smits) [inline ElasticSynchronizer/Helperclasses/JsonParser.cs:7]: This class is not used I think. Can it be removed?

File: ElasticSynchronizer/Program.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        public static IHostBuilder CreateHostBuilder(string[] args) =>
+
+
+            Host.CreateDefaultBuilder(args)
+                .ConfigureServices((hostContext, services) =>
+                {
+                    string environmentName = Environment.GetEnvironmentVariable("ELASTIC_DOTNET_ENVIRONMENT");
+
+                    IConfiguration configuration = new ConfigurationBuilder()
+                                                   .AddJsonFile("appsettings.json", true, true)
```
Comment (Brend-Smits) [inline ElasticSynchronizer/Program.cs:45]: Should this also be renamed into appsettingselastic.json to prevent collisions?

File: Models/Project.cs [hunk 3]
```diff
@@ -61,6 +62,7 @@ public Project()
 
         public CallToAction CallToAction { get; set; }
 
+        [JsonIgnore]
```
Comment (Brend-Smits) [inline Models/Project.cs:65]: Why is this here and what purpose does it serve?

File: NotificationSystem/Properties/PublishProfiles/FolderProfile.pubxml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<!--
+https://go.microsoft.com/fwlink/?LinkID=208121. 
+-->
+<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <Configuration>Release</Configuration>
+    <Platform>Any CPU</Platform>
+    <PublishDir>C:\Users\Max\Documents\PUblished</PublishDir>
```
Comment (Brend-Smits) [inline NotificationSystem/Properties/PublishProfiles/FolderProfile.pubxml:9]: This does not look right. Should this file even be included in the repo?

File: Services/Services/UserService.cs [hunk 5]
```diff
@@ -72,14 +94,59 @@ public bool UserHasScope(string identityId, string scope)
...
+        }
+
+        private IEnumerable<int> GetSimilarUsers(int userId)
+        {
+            List<int> foundUserIds = Repository.GetSimilarUsers(userId);
+
+            // If no similar users were found then throw error.
+            if(foundUserIds.Count == 0)
+            {
+                throw new RecommendationNotFoundException("no similar user(s) were found.");
```
Comment (Brend-Smits) [inline Services/Services/UserService.cs:143]:
```suggestion
                throw new RecommendationNotFoundException("No similar user(s) were found.");
```

File: Services/Services/UserService.cs [hunk 6]
```diff
@@ -72,14 +94,59 @@ public bool UserHasScope(string identityId, string scope)
...
+                    {
+                        return recommendedProjects;
+                    }
+                }
+            }
+
+            // If no recommendations were found for similar users then throw error. 
+            if(recommendedProjects.Count == 0)
+            {
+                throw new RecommendationNotFoundException("similar user(s) do not have any projects you do not like yet.");
```
Comment (Brend-Smits) [inline Services/Services/UserService.cs:130]:
```suggestion
                throw new RecommendationNotFoundException("Similar user(s) do not have any projects you do not like yet.");
```

File: elasticsearch/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,7 @@
+ARG ELK_VERSION
+
+# https://www.docker.elastic.co/
+FROM docker.elastic.co/elasticsearch/elasticsearch:${ELK_VERSION}
+
+# Add your elasticsearch plugins setup here
+# Example: RUN elasticsearch-plugin install analysis-icu
```
Comment (Brend-Smits) [inline elasticsearch/Dockerfile:7]: Do we need this folder? If so, is it only used to host the dockerfile of ES? 
What are the other files used for in the analysis folder?

**Review by Brend-Smits — DISMISSED** [review]
In terms of general performance, how does it affect things like adding a project, searching for a project, liking a project, and things like that? Will it affect performance for day-to-day operations like such?

File: Models/Project.cs (hunk 3 above)
Comment (niraymak) [reply Models/Project.cs:66]: Newtonsoft.Jsonconvert created an infinite loop when we tried to send the project to the messagebroker.

File: elasticsearch/Dockerfile (hunk 7 above)
Comment (niraymak) [reply elasticsearch/Dockerfile:7]: The files in the analysis folder are used to recognize synonyms.
So the folder is used both for the Dockerfile and the synonym config.

**Review by DamienKusters — APPROVED** [review]

**Review by DamienKusters — APPROVED** [review]

**Review by DamienKusters — DISMISSED** [review]

File: API/Controllers/ProjectController.cs [hunk 8]
```diff
@@ -100,6 +100,32 @@ public class ProjectController : ControllerBase
...
+
+        /// <summary>
+        ///     This method is responsible for retrieving all projects in ElasticSearch formatted model. After these project are retrieved the endpoint registers the projects at the messagebroker to synchronize.
+        /// </summary>
+        /// <returns>This method returns a list of in ElasticSearch formatted projects.</returns>
+        /// <response code="200">This endpoint returns a list of in ElasticSearch formatted projects.</response>
+        [HttpGet("export")]
+        [ProducesResponseType(typeof(ProjectResultsResource), (int) HttpStatusCode.OK)]
+        [ProducesResponseType(typeof(ProblemDetails), (int) HttpStatusCode.BadRequest)]
+        public async Task<IActionResult> MigrateDatabaseToElasticSearch()
```
[the remark this answers is not in the document]
Comment (niraymak) [reply API/Controllers/ProjectController.cs:121]: Endpoint has authorization now.

The endpoint deletes the dataset, recreates the dataset and then migrates all the data. So no data will be lost.
RabbitMQ helps not draining the performance of this operation.

**Comment by niraymak** [comment]
> In terms of general performance, how does it affect things like adding a project, searching for a project, liking a project, and things like that? Will it affect performance for day-to-day operations like such?

I compared Postman test results of this branch against another branch. 
Beneath you can find some of the results. Out of these results I can not conclude that there is a huge difference in performance. 

**This branch**
https://github.com/DigitalExcellence/dex-backend/pull/359/checks?check_run_id=2175304446
6m35s for 3 iterations of Postman tests. Amount of tests are equal to the other branch. (we made a new step for recommendation integration tests)

↳ Project-CreateProjectToFollow-Administrator
  POST https://localhost:5005/connect/token [200 OK, 1.13KB, 11ms]

↳ Project-LikeProject-Administrator
  POST https://localhost:5005/connect/token [200 OK, 1.13KB, 13ms]

↳ Search-SearchInternal-Administrator
  POST https://localhost:5005/connect/token [200 OK, 1.13KB, 8ms]

↳ Project-DeleteProject-Administrator
  POST https://localhost:5005/connect/token [200 OK, 1.13KB, 7ms]

**Non recommendation system branch**
https://github.com/DigitalExcellence/dex-backend/pull/408/checks?check_run_id=2174364502
6m15s for 3 iterations of Postman tests.

↳ Project-CreateProjectToFollow-Administrator
  POST https://localhost:5005/connect/token [200 OK, 1.13KB, 16ms]

↳ Project-LikeProject-Administrator
  POST https://localhost:5005/connect/token [200 OK, 1.13KB, 8ms]

↳ Search-SearchInternal-Administrator
  POST https://localhost:5005/connect/token [200 OK, 1.13KB, 14ms]

↳ Project-DeleteProject-Administrator
  POST https://localhost:5005/connect/token [200 OK, 1.13KB, 6ms]

File: .github/workflows/run-tests.yml [hunk 9]
```diff
@@ -37,10 +37,10 @@ jobs:
     - name: Docker Compose
       run: docker-compose up --build -d
 
-    - name: Sleep for 30 seconds
+    - name: Sleep for 90 seconds
       uses: juliangruber/sleep-action@v1
       with:
-        time: 30s
+        time: 90s
```
Comment (JVerbruggen) [inline .github/workflows/run-tests.yml:43]: Whats this used for?

File: NotificationSystem/Services/EmailSender.cs [hunk 10]
```diff
@@ -43,21 +44,19 @@ public EmailSender(ISendGridClient sendGridClient, string emailFrom, bool sandbo
 
         public void ParsePayload(string jsonBody)
         {
+            notification = JsonConvert.DeserializeObject<EmailNotification>(jsonBody);
             Notification = JsonConvert.DeserializeObject<EmailNotification>(jsonBody);
```
Comment (JVerbruggen) [inline NotificationSystem/Services/EmailSender.cs:48]: why use 2 similar variables?

Comment (JVerbruggen) [reply NotificationSystem/Services/EmailSender.cs:48]: Also the capitalized one is not used

File: Repositories.Tests/DataGenerators/ElasticSearchResults.cs [hunk 11]
```diff
@@ -0,0 +1,17 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Repositories.Tests.DataGenerators
+{
+    public class ElasticSearchResults
```
Comment (JVerbruggen) [inline Repositories.Tests/DataGenerators/ElasticSearchResults.cs:7]: What.. is happening here? Please add summaries

File: Repositories/ElasticSearch/ProjectConverter.cs [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+                foreach(ProjectLike projectLike in project.Likes)
+                {
+                    likes.Add(projectLike.UserId);
+                }
+            }            
+            convertedProject.Description = project.Description;
+            convertedProject.ProjectName = project.Name;
+            convertedProject.Id = project.Id;
+            convertedProject.Created = project.Created;
+            convertedProject.Likes = likes;
```
Comment (JVerbruggen) [inline Repositories/ElasticSearch/ProjectConverter.cs:39]: This is kind of a big nope I think. You're using the reference now to edit the project, and you don't return anything. To improve readability, try rewriting this so that it uses a return statement.

File: docker-compose-production.yml [hunk 13]
```diff
@@ -163,7 +167,39 @@ services:
         - Config__RabbitMQ__Password=${App__RabbitMQ__Password}
     networks:
         rabbitmq-network:
```
Comment (JVerbruggen) [inline docker-compose-production.yml:169]: this one doesnt use mssql network anymore? Or did you remove this by accident?

File: API/HelperClasses/EmailSender.cs [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+        /// <param name="recipient">The email address of the recipient</param>
+        /// <param name="textContent">The text content of the email</param>
+        /// <param name="htmlContent">The HTML content of the email</param>
+        public void Send(string recipient, string textContent, string htmlContent = null)
+        {
+            EmailNotificationRegister emailNotification = new EmailNotificationRegister(recipient, textContent, htmlContent);
+            taskPublisher.RegisterTask(Newtonsoft.Json.JsonConvert.SerializeObject(emailNotification), Subject.EMAIL);
+        }
+    }
+}
```
Comment (JVerbruggen) [inline API/HelperClasses/EmailSender.cs:48]: Duplicate from the notificationsystem one? Maybe add this to model layer? Or nah?

**Review by JVerbruggen — CHANGES_REQUESTED** [review]
Impressive work. I requested several changes but I don't think they're too big of a deal.

File: .github/workflows/run-tests.yml (hunk 9 above)
Comment (niraymak) [reply .github/workflows/run-tests.yml:43]: Changed because 2 extra projects are started. Earlier when the sleep was still 30s the Postman test seemed to start when applications were not started yet.
After changing to 90s it worked immediately, so that's why we changed it.

File: API/1_API.csproj [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+  <ItemGroup>
+    <ProjectReference Include="..\Data\4_Data.csproj" />
+    <ProjectReference Include="..\MessageBrokerPublisher\MessageBrokerPublisher.csproj" />
+    <ProjectReference Include="..\Models\5_Models.csproj" />
+    <ProjectReference Include="..\NotificationSystem\NotificationSystem.csproj" />
+    <ProjectReference Include="..\Services\2_Services.csproj" />
+  </ItemGroup>
+
+
+</Project>
```
[the remark this answers is not in the document]
Comment (niraymak) [reply API/1_API.csproj]: Nicely spotted. This was a mistake after merging develop. Removed redundant .csproj files.

File: API/01_API.csproj [hunk 16]
```diff
@@ -41,18 +41,17 @@
...
         <PackageReference Include="Swashbuckle.AspNetCore" Version="5.0.0" />
-        <PackageReference Include="Serilog.AspNetCore" Version="3.2.0" />
+        <PackageReference Include="Serilog.AspNetCore" Version="3.4.0" />
     </ItemGroup>
 
     <ItemGroup>
         <ProjectReference Include="..\Data\06_Data.csproj" />
         <ProjectReference Include="..\MessageBrokerPublisher\09_MessageBrokerPublisher.csproj" />
         <ProjectReference Include="..\Models\07_Models.csproj" />
-        <ProjectReference Include="..\NotificationSystem\11_NotificationSystem.csproj" />
```
[the remark this answers is not in the document]
Comment (Maxvanhattum) [reply API/01_API.csproj:55]: This dependency snuck in from some reason, basically the API should only depend on the MessageBrokerPublisher, this because we want the notificationsystem to be decoupled from the API. So the API should publish a task to the message broker, then the notification system gets notified by the messagebroker and does its job. 
This way we could deploy the notificationsystem if necessary on a different server.

File: API/Controllers/UserController.cs [hunk 17]
```diff
@@ -594,6 +597,44 @@ public async Task<IActionResult> SetUserGraduationDate([FromBody] UserResource u
             return Ok(mapper.Map<User, UserResourceResult>(user));
         }
 
+
+        /// <summary>
+        /// Get recommended projects for the user who is logged in.
+        /// </summary>
+        /// <returns> Ok </returns>
+        [HttpGet("projectrecommendations/{amount}")]
+        public async Task<IActionResult> GetRecommendedProjects(int amount)
```
[the remark this answers is not in the document]
Comment (niraymak) [reply API/Controllers/UserController.cs:608]: Added Authorize attribute. :)

File: API/Controllers/UserController.cs [hunk 18]
```diff
@@ -594,6 +597,44 @@ public async Task<IActionResult> SetUserGraduationDate([FromBody] UserResource u
             return Ok(mapper.Map<User, UserResourceResult>(user));
         }
 
+
+        /// <summary>
+        /// Get recommended projects for the user who is logged in.
+        /// </summary>
+        /// <returns> Ok </returns>
+        [HttpGet("projectrecommendations/{amount}")]
```
[the remark this answers is not in the document]
Comment (niraymak) [reply API/Controllers/UserController.cs:605]: Done!

File: Data/4_Data.csproj [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+
+  <ItemGroup>
+    <ProjectReference Include="..\Models\5_Models.csproj" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <Folder Include="Configurations\IdentityMigrations\" />
+  </ItemGroup>
+
+</Project>
```
[the remark this answers is not in the document]
Comment (niraymak) [reply Data/4_Data.csproj]: Removed it.

File: ElasticSynchronizer/Workers/UpdateProjectWorker.cs [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+
+            ICallbackService documentUpdaterService = new DocumentUpdater(config, restClient);
+            EventingBasicConsumer consumer = listener.CreateConsumer(documentUpdaterService);
+
+            listener.StartConsumer(consumer, subject);
+        }
+
+    }
+
+}
```
[the remark this answers is not in the document]
Comment (Maxvanhattum) [reply ElasticSynchronizer/Workers/UpdateProjectWorker.cs:66]: We too had a discussion about this, like asking for the subject and callback object in the constructor, however we ran into the problem that we can't pass along parameters into the services.AddHostedService<DeleteProjectWorker>();
Moreover we also argued that while the code looks the same, it does something else, hence it is not repeating code.
However if you do find or have a solution for this problem, we are open for suggestions.

File: NotificationSystem/Services/EmailSender.cs (hunk 10 above)
Comment (Maxvanhattum) [reply NotificationSystem/Services/EmailSender.cs:48]: Basically the Notification was used to check the value while testing, but the way it is coded now is indeed not the right way. We've encapsulated the field now with a getter.

File: docker-compose-production.yml (hunk 13 above)
Comment (niraymak) [reply docker-compose-production.yml:169]: Added again. Thanks for mentioning.

File: API/HelperClasses/EmailSender.cs (hunk 14 above)
Comment (Maxvanhattum) [reply API/HelperClasses/EmailSender.cs:48]: They have the same name but do two different things, the helper class registers the 'send email' task to the message broker. 
The emailsender in the notification system is a class that gets instantiated to be passed into the message broker consumer so that it can be used as a callback.
To reiterate, we want the notification system (and elasticsynchronizer/jobscheduler) to be separate services that can be ran independently but loosely coupled.

File: docker-compose-production.yml (hunk 13 above)
Comment (niraymak) [reply docker-compose-production.yml:169]: In the future we should change this as the jobscheduler is now communicating internally, this should be externally. But this is out of scope of this PR.

**Review by DamienKusters — DISMISSED** [review]

**Review by JVerbruggen — APPROVED** [review]

**Review by DamienKusters — APPROVED** [review]
